Fix #122 by enabling joints to connect rods and rigid bodies

mstoelzle opened the pull request:
## Features and Changes
This issue fixes #122 by implementing the following changes and features:

1. Enables the connection of Cosserat rods and rigid bodies in any arbitrary combination for the following joint types: `FreeJoint`, `HingeJoint`, and `FixedJoint`
2. The `FreeJoint` constrains the translational movement of both systems at the specified connection point. As the point of joint connection doesn't necessarily coincide with the center of mass of the rigid body or the node of the rod, the user can use the `point_system_one` and `point_system_two` parameters to specify how the connection point relates to the node / CoM in the local frame of the system.
3. For the `HingeJoint`,  the `link_direction` of the second system is now extracted from the `director_collection` instead of subtracting node positions
4. The `FixedJoint` class now also accepts the `point_system_one` and `point_system_two` parameters and forwards them to the parent class `FreeJoint`

## Examples of Usage
How to connect two Cosserat rods together using a spherical joint with a gap of 0.01 m in between.

```py
simulator.connect(rod_one, rod_two).using(
    FreeJoint,
    k=1e4,
    nu=1,
    point_system_one=np.array([0.0, 0.0, 0.005]),
    point_system_two=np.array([0.0, 0.0, -0.005]),
)
```
    
How to connect the distal end of a CosseratRod with the base of a cylinder using a spherical joint.

```py
simulator.connect(rod, cylinder).using(
    FreeJoint,
    k=1e4,
    nu=1,
    point_system_two=np.array([0.0, 0.0, -cylinder.length / 2.]),
)
```

How to connect a cosserat rod with the base of a cylinder using a fixed joint, where the cylinder is rotated by 45 degrees around the y-axis.

```py
from scipy.spatial.transform import Rotation
simulator.connect(rod, cylinder).using(
      FixedJoint,
      k=1e5,
      nu=1e0,
      kt=1e3,
      nut=1e-3,
      point_system_two=np.array([0, 0, -cylinder.length / 2]),
      rest_rotation_matrix=Rotation.from_euler('y', np.pi / 4, degrees=False).as_matrix(),
)
```

## Results from `JointCases`
**Spherical joint:** 

System consisting of two rods and one cylinder connected with spherical joints. The sinusoidal force is applied to the second rod. Example can be run using:

```bash
python examples/JointCases/spherical_joint.py
```

https://user-images.githubusercontent.com/2360366/180027045-d9753a1a-56ec-4459-a719-a13c7398620e.mp4

**Fixed joint:** 

System consisting of two rods and one cylinder connected with fixed joints. The sinusoidal force is applied to the second rod. Example can be run using:

```bash
python examples/JointCases/fixed_joint.py
```

https://user-images.githubusercontent.com/2360366/180449881-c806dd79-324b-40d9-9f57-b6257c7f8978.mp4

## TODOs

- [ ] Write tests for connecting rods with cylinders, particularly for non-zero `point_system_one` and `point_system_two` parameters
- [x] Add cylinder to `spherical_joint.py` example as third element.
- [x] Add cylinder to `fixed_joint.py` example as third element

armantekinalp commented:
@mstoelzle is the simulation in second video explodes ?

armantekinalp commented on elastica/rod/cosserat_rod.py:
```diff
@@ -520,6 +520,164 @@ def compute_shear_energy(self):
+        >>> rod.compute_position_of_point(np.array([0, 0, 1]), -1)
+        """
+        assert point.shape == (3,), "Point must be in the shape (3,)"
+        position = self.position_collection[..., index] + np.dot(
```
I still need to look PR, but why do you define the point as an actual point in space instead of some percentage of lengths. Since rod element lengths can change due to extension.

mstoelzle replied:
@armantekinalp Thanks for your question. These points are always in the local coordinate system relative to a node. So when the rod elongates, this point moves with the node. The main idea for usage is for the ends of the rods: the joint might not be attached along the backbone of the rod, but rather touching the cylindrical surface. Or the joint might be displaced by a small distance from the end of a rod through an (unmodelled) rigid element.

bhosale2 replied:
@armantekinalp  thoughts?

bhosale2 commented:
@mstoelzle we can focus on this one after #143 is completed, which is closer to completion.

bhosale2 commented:
@mstoelzle what's the status here? The PR says WIP do you still need to implement something?

mstoelzle commented:
> @mstoelzle what's the status here? The PR says WIP do you still need to implement something?

@bhosale2 Although some tests are still missing, I would like to ask you to already have a look at this PR. I would like to prevent writing some tests for functionality / interfaces we might not need / do differently at the end.

bhosale2 commented:
@mstoelzle I went through the PR briefly and noticed that 2 concepts are being implemented simultaneously here:

1. Joint between the rod and rigid body
2. The joint itself is now configurable via the `point_system` feature that you introduce here, where the entities can now be connected at different points, and not constrained to the end of the rigid body or the node of the Cosserat rod element.

In my opinion, the second feature shows some base class structure change, and also should be a separate PR in itself? Because the connection point feature also adds more configurability to joints between rods themselves. Maybe here we stick to the more frequent version of connections at ends?

Let's take a call after seeing what @skim0119 and @armantekinalp think.

mstoelzle commented:
> @mstoelzle I went through the PR briefly and noticed that 2 concepts are being implemented simultaneously here:
> 
> 
> 
> 1. Joint between the rod and rigid body
> 
> 2. The joint itself is now configurable via the `point_system` feature that you introduce here, where the entities can now be connected at different points, and not constrained to the end of the rigid body or the node of the Cosserat rod element.
> 
> 
> 
> In my opinion, the second feature shows some base class structure change, and also should be a separate PR in itself? Because the connection point feature also adds more configurability to joints between rods themselves. Maybe here we stick to the more frequent version of connections at ends?
> 
> 
> 
> Let's take a call after seeing what @skim0119 and @armantekinalp think.

Thanks @bhosale2 for having a look at the PR. I agree with your analysis that this PR incorporates the two components you mentioned. It is important however to keep in mind the following points:

1. If we want to connect a rod with a cylinder at the nodes (e.g. center of mass of the cylinder - so the traditional way), there is actually only the tiny change in the 'HingeJoint' class necessary. The 'FreeJoint' wouldn't need any modifications.
2. However, while connecting at a node is suitable for Cosserat rods most of the time, I think it is not useful for rigid bodies in a most cases, as we would hardly want the Center of Mass of the Rigid Body to coincide with the node of the Cosserat Rod. The much larger impact can be created when we give the user the design freedom to choose at which point the joint should be attached to the rigid body: for example the bottom or top of a cylinder etc. 
3. Even if we come up with some hard-coded strategy such as that cylinders are always connected at its base at radius zero (e.g. end of cylinder), most of the code changes would still be necessary and the PR would only be slightly smaller.

armantekinalp commented on elastica/rod/cosserat_rod.py:
```diff
@@ -520,6 +520,164 @@ def compute_shear_energy(self):
             * (_batch_dot(sigma_diff, shear_internal_forces) * self.rest_lengths).sum()
         )
 
+    def compute_position_of_point(self, point, index):
```
Let's remove `compute_position_of_point` and  `compute_velocity_of_point` from Cosserat Rod and include them in joint file as functions.

mstoelzle replied:
Implemented in https://github.com/GazzolaLab/PyElastica/pull/149/commits/74a9c9d3f1511d9aed73a7f7f7b804cf0bebb284

armantekinalp commented on elastica/rigidbody/rigid_body.py:
```diff
@@ -104,3 +104,76 @@ def compute_rotational_energy(self):
         )
         return 0.5 * np.einsum("ik,ik->k", self.omega_collection, J_omega).sum()
+
+    def compute_position_of_point(self, point, index: int = 0):
```
Let's remove `compute_position_of_point` and  `compute_velocity_of_point` from RigidBody and include them in joint file as functions.

mstoelzle replied:
Implemented in https://github.com/GazzolaLab/PyElastica/pull/149/commits/74a9c9d3f1511d9aed73a7f7f7b804cf0bebb284

armantekinalp reviewed (changes requested):
1. Lets keep the previous `FreeJoint` implementation as it is and introduce a new class to connect rigid body and Cosserat Rod.
2. I understand why you need to connect cylinder at some other point then the COM, but what is the reason to connect rod a point different than its nodes? Since always discretization of the rod can be changed. 
3. Also please remove `compute_position_of_point` method from Cosserat Rod and Rigid body and include inside the joint.py file as a free function.

mstoelzle commented:
> * Lets keep the previous `FreeJoint` implementation as it is and introduce a new class to connect rigid body and Cosserat Rod.

As `FixedJoint` is inheriting from `FreeJoint` and also needs to work with rigid bodies, this would require us to do a separate `FixedJoint` class as well to connect rigid bodies. 

- Does this make sense to have this much duplicate code? 
- How do you propose to call the `FreeJoint` and `FixedJoint`, which are able to connect to rigid bodies?

> * I understand why you need to connect cylinder at some other point then the COM, but what is the reason to connect rod a point different than its nodes? Since always discretization of the rod can be changed.

Thanks for your comment. I would like to divide my response into three parts:

- Allowing points as input parameters for both rigid bodies and Cosserat rod nodes makes the implementation easier, the interface cleaner and requires less _if_ statements, as we do not need to distinguish between system types / classes.
- While `point` offsets in z-dir from the Cosserat rod joint can be compensated by choosing a different discretization, as you propose, any radial offset cannot be evoked by choosing a different Cosserat rod configuration, as the mass distribution would be different.
- I currently do not have a specific example / application in mind for defining a point relative to a node. Still, I think it's nice to give the user the freedom to tailor the joint to their use-case, also with regard to connecting Cosserat rods

> * Also please remove `compute_position_of_point` method from Cosserat Rod and Rigid body and include inside the joint.py file as a free function.

Thanks - I will do that.

skim0119 reviewed (commented):
@mstoelzle Thank you for your work! I liked your idea and implementation in general :)

I appreciate your effort to make the joint class more generalizable for both rod and rigid body base types. If you want to complete this PR within v0.3, I have a couple of suggestions:

- Lets move `point-system` implementation to the `experimental` folder.
  - Create `experimental/connection_contact_joint/rod_rigidbody_connection.py`
  - We are currently working on refactoring joint class and others (#113), and we don't want to directly merge any code into `joint.py` until it is sorted out.
- If you have any paper or research experiment to validate, please consider sharing it with us. We want to have some confidence in our simulation regarding physical validation.
- In the future with more data, we'll refactor them into the main `joint.py` module.

Other code styles and documentation look good.

@armantekinalp @bhosale2 Thoughts?

armantekinalp commented:
I think moving this implementation under experiments will help us doing refactoring later on. Similarly, we can have rigid body rod connection in different examples, so future users will have more references to look. So addition to the @skim0119  comments, I would suggest renaming the rigid-body rod connection examples for spherical and fixed joint as, `rod_cylinder_spherical_joint.py` and `rod_cylinder_fixed_joint.py` and have them under different folder.

mstoelzle commented:
> @mstoelzle Thank you for your work! I liked your idea and implementation in general :)

Thanks everybody for all your help in reviewing and suggesting changes for all the PRs in the last few weeks. I appreciate it and I think it definitely improved the code quality!

> 
> * Lets move `point-system` implementation to the `experimental` folder.
>   
>   * Create `experimental/connection_contact_joint/rod_rigidbody_connection.py`

@skim0119 I have done that in https://github.com/GazzolaLab/PyElastica/pull/149/commits/ff77782ce9b9ffb0c0cf690ce45fdec3d472565b


> So addition to the @skim0119 comments, I would suggest renaming the rigid-body rod connection examples for spherical and fixed joint as, `rod_cylinder_spherical_joint.py` and `rod_cylinder_fixed_joint.py` and have them under different folder.

@armantekinalp I have addressed this in https://github.com/GazzolaLab/PyElastica/pull/149/commits/ff77782ce9b9ffb0c0cf690ce45fdec3d472565b as well

bhosale2 commented on elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py:
```diff
@@ -0,0 +1,417 @@
+
+class RodRigidBodyFreeJoint(FreeJoint):
+    """
+    This free joint class is the base class for all joints. Free or spherical
```
```suggestion
    This free joint class is the base class for all joints between a rod and rigid body. Free or spherical
```

bhosale2 commented on elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py:
```diff
@@ -0,0 +1,417 @@
+        Stiffness coefficient of the joint.
+    nu: float
+        Damping coefficient of the joint.
+    point_system_one : numpy.ndarray
+            Describes for system one in the local coordinate system the translation from the node `index_one` (for rods)
+            or the center of mass (for rigid bodies) to the joint.
+    point_system_two : numpy.ndarray
+        Describes for system two in the local coordinate system the translation from the node `index_two` (for rods)
+        or the center of mass (for rigid bodies) to the joint.
```
shouldn't one of these refer to the rod and the other to the rigid body now? based on the naming?

mstoelzle replied:
@bhosale2 I have a different opinion than you on this one. I want to leave the option open for the user to combine systems in any way they would like:

- **Rod to Rod:** In this situation, the users can also use the standard (non-experimental) `FreeJoint` implementation, which should numerically give the same result, but is possibly computationally slightly faster
- **Rod to Rigid Body**
- **Rigid Body to Rod**: This should have the same exact effect as the second bullet point, provided that the user mirrors the configuration.
- **Rigid Body to Rigid Body**: Most importantly, this class should also enable this configuration. If I force the first system to be a rod and the second system to be a rigid body, this is not possible anymore and not implemented anywhere else

I am open though to changing the name of the file / of the classes if you think something would be more intuitive than the current naming scheme.

bhosale2 replied:
I appreciate the generality seen in the code here, but it is not consistent with the naming of the function then, which when seen one expects a joint between a rod and rigid body. In that case something like a generic_joint would make more sense? generalised_joint? @armantekinalp can voice his opinion as well on this

mstoelzle replied:
@bhosale2 I see those two classes as a more general implementation of the original FreeJoint and FixedJoint, which should be called something like Rod2RodFreeJoint and Rod2RodFixedJoint in the future. 

So I would be actually in favor in calling the new, more general implementation either FreeJoint and FixedJoint or NewFreeJoint and  NewFixedJoint to communicate to the user that this will be (most likely) the future standard implementation of these joint classes.

bhosale2 replied:
I agree with the first part, but refactoring the core codebase `FreeJoint` has its own drawbacks. `NewFreejoint` may be a good idea. Anyways, for this PR we are shifting these to experimental as planned, and we can revisit this naming then later in the future.

mstoelzle replied:
@armantekinalp If you agree as well, I will rename it accordingly to NewFreeJoint and NewFixedJoint.  They are in the experimental module anyway.

armantekinalp replied:
Hi @mstoelzle . After discussing with @bhosale2 and @skim0119,  we think we can rename them as `GenericSystemTypeFixedJoint` and `GenericSystemTypeFreeJoint`for this PR.

mstoelzle replied:
Thanks for this suggestion. I renamed in https://github.com/GazzolaLab/PyElastica/pull/149/commits/002770e757697ff071a419358aa75af4d0b9839c

bhosale2 commented on elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py:
```diff
@@ -0,0 +1,417 @@
+
+    Examples
+    --------
+    How to connect two Cosserat rods together using a spherical joint with a gap of 0.01 m in between.
+
+    >>> simulator.connect(rod_one, rod_two, first_connect_idx=-1, second_connect_idx=0).using(
+    ...    FreeJoint,
+    ...    k=1e4,
+    ...    nu=1,
+    ...    point_system_one=np.array([0.0, 0.0, 0.005]),
+    ...    point_system_two=np.array([0.0, 0.0, -0.005]),
+    ... )
+
+    How to connect the distal end of a CosseratRod with the base of a cylinder using a spherical joint.
+
+    >>> simulator.connect(rod, cylinder, first_connect_idx=-1, second_connect_idx=0).using(
+    ...    FreeJoint,
+    ...    k=1e4,
+    ...    nu=1,
+    ...    point_system_two=np.array([0.0, 0.0, -cylinder.length / 2.]),
+    ... )
+
+    """
```
modify to reflect joint functionality

bhosale2 commented on elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py:
```diff
@@ -0,0 +1,417 @@
+           Stiffness coefficient of the joint.
+        nu: float
+           Damping coefficient of the joint.
+        point_system_one : numpy.ndarray
+            Describes for system one in the local coordinate system the translation from the node `index_one` (for rods)
+            or the center of mass (for rigid bodies) to the joint.
+            (default = np.array([0.0, 0.0, 0.0]))
+        point_system_two : numpy.ndarray
+            Describes for system two in the local coordinate system the translation from the node `index_two` (for rods)
+            or the center of mass (for rigid bodies) to the joint.
+            (default = np.array([0.0, 0.0, 0.0]))
```
^ditto

bhosale2 commented on elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py:
```diff
@@ -0,0 +1,417 @@
+
+        Parameters
+        ----------
+        system_one : SystemType
+             System two of the joint connection.
+             Object of a child class of  either `CosseratRod` or `RigidBodyBase`.
+        index_one : int
+            Index of first system for joint.
+        system_two : SystemType
+            System two of the joint connection.
+            Object of a child class of  either `CosseratRod` or `RigidBodyBase`.
+        index_two : int
+            Index of second system for joint.
+
```
if we know the joint is between a rod and a rigid body maybe like the external contact we can ask for the first body to be a rod and the second to be a rigid body? modify docstring and functionality accordingly?

bhosale2 commented on elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py:
```diff
@@ -0,0 +1,417 @@
+        # Compute the velocity deviation of the point belonging to system one from the point belonging to system two
+        relative_velocity = velocity_system_two - velocity_system_one
+        # only penalize (e.g. apply opposite damping force) velocity aligned with the `normalized_end_distance_vector`
+        normal_relative_velocity = (
+            np.dot(relative_velocity, normalized_end_distance_vector)
+            * normalized_end_distance_vector
+        )
```
I don't think we need this projection, we can directly have a resistive force along the direction of relative velocity

bhosale2 replied:
@mstoelzle  updates on these?

mstoelzle replied:
@bhosale2 Sorry, I missed this comment. GitHub issues are sometimes a bit messy...

I think I am the wrong person to ask about this. I literally copied this from the original 'FreeJoint' implementation. 

Out of my intuition, I don't think this projection is necessary. Please let me know if I can remove it.

bhosale2 replied:
I just confirmed with @armantekinalp, you can go ahead.

bhosale2 commented on elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py:
```diff
@@ -0,0 +1,417 @@
+        # Below if check is not efficient find something else
+        # We are checking if the point of `system_one` and point of `system_two` are at the same point in space
+        # If they are at the same point in space, it is a zero vector.
+        if end_distance <= Tolerance.atol():
+            normalized_end_distance_vector = np.array([0.0, 0.0, 0.0])
+        else:
+            normalized_end_distance_vector = end_distance_vector / end_distance
```
this wont be needed based on the comment below

bhosale2 replied:
updates here @mstoelzle ?

bhosale2 commented on elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py:
```diff
@@ -0,0 +1,417 @@
+
+        Parameters
+        ----------
+        system_one : SystemType
+             System two of the joint connection.
+             Object of a child class of  either `CosseratRod` or `RigidBodyBase`.
+        index_one : int
+            Index of first system for joint.
+        system_two : SystemType
+            System two of the joint connection.
+            Object of a child class of  either `CosseratRod` or `RigidBodyBase`.
+        index_two : int
+            Index of second system for joint.
```
modify as mentioned above

bhosale2 commented on elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py:
```diff
@@ -0,0 +1,417 @@
+            Rotational stiffness coefficient of the joint.
+        nut: float
+            Rotational damping coefficient of the joint.
+        point_system_one : numpy.ndarray
+            Describes for system one in the local coordinate system the translation from the node `index_one` (for rods)
+            or the center of mass (for rigid bodies) to the joint.
+        point_system_two : numpy.ndarray
+            Describes for system two in the local coordinate system the translation from the node `index_two` (for rods)
+            or the center of mass (for rigid bodies) to the joint.
```
reflect as above

bhosale2 commented on elastica/experimental/connection_contact_joint/rod_rigidbody_connection.py:
```diff
@@ -0,0 +1,417 @@
+            Rotational stiffness coefficient of the joint.
+        nut: float = 0.
+            Rotational damping coefficient of the joint.
+        point_system_one : numpy.ndarray
+            Describes for system one in the local coordinate system the translation from the node `index_one` (for rods)
+            or the center of mass (for rigid bodies) to the joint.
+            (default = np.array([0.0, 0.0, 0.0]))
+        point_system_two : numpy.ndarray
+            Describes for system two in the local coordinate system the translation from the node `index_two` (for rods)
+            or the center of mass (for rigid bodies) to the joint.
+            (default = np.array([0.0, 0.0, 0.0]))
```
^ditto

bhosale2 reviewed (changes requested):
Minor comments, nice work overall!

armantekinalp reviewed (changes requested):
Minor changes.

mstoelzle commented on tests/test_joint.py:
```diff
@@ -207,12 +221,9 @@ def test_hingejoint():
         rod2.external_forces[..., rod2_index], -1 * contactforce, atol=Tolerance.atol()
     )
 
-    linkdirection = (
-        rod2.position_collection[..., rod2_index + 1]
-        - rod2.position_collection[..., rod2_index]
-    )
```
@armantekinalp I changed the computation of the link direction so that it is compatible with rigid bodies. The computation using the position of multiple nodes only works for rods.

mstoelzle commented on examples/JointCases/fixed_joint.py:
```diff
@@ -75,7 +81,13 @@ class FixedJointSimulator(
+    FixedJoint,
+    k=1e5,
+    nu=0e0,
+    kt=1e1,
```
@armantekinalp I actually badly tuned this spring constant for the torsion before in the previous PR. The example simulation was very unstable and fragile as it would disconverge often. 

As this spring is now directly acting on the rotation instead some position difference (as changed in previous PR), a lower spring constant seems to be suitable.

armantekinalp replied:
okay thanks for clarifying.

armantekinalp reviewed (approved):
LGTM

bhosale2 commented on docs/api/connections.rst:
```diff
@@ -26,8 +26,10 @@ Compatibility
 Connection / Contact / Joints   Rod  Rigid Body
 =============================== ==== =========== 
 FreeJoint                       ✅   ❌
+GenericSystemTypeFreeJoint      ✅   ✅
 ExternalContact                 ✅   ❌
 FixedJoint                      ✅   ❌
+GenericSystemTypeFixedJoint     ✅   ✅
```
please remove from docs, we don't host experimental features here yet.

mstoelzle replied:
Ok, I see. I removed it in https://github.com/GazzolaLab/PyElastica/pull/149/commits/63be1a41d9aac93826a52d3171766d4984ab2d45

bhosale2 commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,414 @@
+        # this implementation timed: 2.48 µs ± 126 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
+        end_distance = np.sqrt(np.dot(end_distance_vector, end_distance_vector))
+
+        # Below if check is not efficient find something else
+        # We are checking if the point of `system_one` and point of `system_two` are at the same point in space
+        # If they are at the same point in space, it is a zero vector.
+        if end_distance <= Tolerance.atol():
+            normalized_end_distance_vector = np.array([0.0, 0.0, 0.0])
+        else:
+            normalized_end_distance_vector = end_distance_vector / end_distance
+
```
```suggestion
```

mstoelzle replied:
Implemented in https://github.com/GazzolaLab/PyElastica/pull/149/commits/cc2010ddebfff9cf7128930dfb6e6e9ad92eb167

bhosale2 commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,414 @@
+
+        # Compute the velocity deviation of the point belonging to system one from the point belonging to system two
+        relative_velocity = velocity_system_two - velocity_system_one
+        # only penalize (e.g. apply opposite damping force) velocity aligned with the `normalized_end_distance_vector`
+        normal_relative_velocity = (
+            np.dot(relative_velocity, normalized_end_distance_vector)
+            * normalized_end_distance_vector
+        )
+        damping_force = -self.nu * normal_relative_velocity
```
```suggestion
        damping_force = -self.nu * relative_velocity
```

mstoelzle replied:
Implemented in https://github.com/GazzolaLab/PyElastica/pull/149/commits/cc2010ddebfff9cf7128930dfb6e6e9ad92eb167

bhosale2 commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,414 @@
+
+    >>> system.compute_position_of_point(np.array([0, 0, 1]), -1)
+    """
+    assert point.shape == (3,), "Point must be in the shape (3,)"
```
Maybe not a good idea to have an `assert` in a function that will be called every time step?

mstoelzle replied:
Removed in https://github.com/GazzolaLab/PyElastica/pull/149/commits/a01136f57363afdc4b835ad163283cbd8926b343

bhosale2 commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,414 @@
+
+    >>> system.compute_velocity_of_point(np.array([0, 0, 1]), -1)
+    """
+    assert point.shape == (3,), "Point must be in the shape (3,)"
```
^ditto

mstoelzle replied:
Done in https://github.com/GazzolaLab/PyElastica/pull/149/commits/a01136f57363afdc4b835ad163283cbd8926b343

bhosale2 reviewed (changes requested):
Minor stuff, good work @mstoelzle

mstoelzle commented:
> 



> Minor stuff, good work @mstoelzle, after this round it will be LGTM from my side.
> Also based on #163, please pull latest `update`, then modify the rod damper class name `ExponentialDamper` to `AnalyticalLinearDamper`.

Thanks for your suggestions. I addressed all of them and I also fixed a bug in the changes for the `HingeJoint`

bhosale2 commented on examples/JointCases/experimental/generic_system_type_fixed_joint.py:
```diff
@@ -0,0 +1,205 @@
+__doc__ = """Fixed joint example, for detailed explanation refer to Zhang et. al. Nature Comm.  methods section."""
+
+import numpy as np
+import sys
+
+# FIXME without appending sys.path make it more generic
+sys.path.append("../../")
```
```suggestion
```

bhosale2 commented on examples/JointCases/experimental/generic_system_type_spherical_joint.py:
```diff
@@ -0,0 +1,201 @@
+methods section."""
+
+import numpy as np
+import sys
+
+# FIXME without appending sys.path make it more generic
+sys.path.append("../../")
```
```suggestion
```

bhosale2 reviewed (changes requested):
Super minor stuff.

bhosale2 reviewed (approved):
LGTM. @skim0119 please take over.

skim0119 commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,396 @@
+        ----------
+        system_one : SystemType
+             System two of the joint connection.
+             Object of a child class of  either `CosseratRod` or `RigidBodyBase`.
```
```suggestion
```

skim0119 commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,396 @@
+            Index of first system for joint.
+        system_two : SystemType
+            System two of the joint connection.
+            Object of a child class of  either `CosseratRod` or `RigidBodyBase`.
```
```suggestion
```

skim0119 reviewed (changes requested):
@mstoelzle LGTM. Minor changes for documentation consistency.

@armantekinalp Do you think we should make a `nightly` doc page that includes experimental modules?

mstoelzle commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,396 @@
+            Describes for system two in the local coordinate system the translation from the node `index_two` (for rods)
+            or the center of mass (for rigid bodies) to the joint.
+            (default = np.array([0.0, 0.0, 0.0]))
+        """
```
Addressed in https://github.com/GazzolaLab/PyElastica/pull/149/commits/f5f4d683adb72aff7c13bd30f4a07398f75ba74b

mstoelzle commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,396 @@
+        ----------
+        system_one : SystemType
+             System two of the joint connection.
+             Object of a child class of  either `CosseratRod` or `RigidBodyBase`.
```
Sorry, I am not quite sure what you mean here. Could you please go into more detail? Possibly it is already resolved by the latest changes

mstoelzle replied:
Fixed in https://github.com/GazzolaLab/PyElastica/pull/149/commits/f907e1ad86eb30b9a3bb2c8affee4d15ac2fc149

mstoelzle commented on tests/test_generic_system_type_joint.py:
```diff
@@ -0,0 +1,277 @@
+__doc__ = """ Joints for generic system types test module """
```
Done in https://github.com/GazzolaLab/PyElastica/pull/149/commits/d64072e11f24a365febbd26c474fa1fdd6afc0c3

mstoelzle commented on tests/test_generic_system_type_joint.py:
```diff
@@ -0,0 +1,277 @@
+from typing import *
+
+# seed random number generator
+np.random.seed(0)
```
Done in https://github.com/GazzolaLab/PyElastica/pull/149/commits/ef5d5c1ae7e0598a791e650c584d95554a9c42a7

mstoelzle commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,396 @@
+        )
+
+    def apply_torques(self, system_one, index_one, system_two, index_two):
+        return FixedJoint.apply_torques(
```
`super()` wouldn't work here, as I need to access the `apply_torques` method of the `FixedJoint` class and I need to specify that I do not want to access the `apply_torques` method of the `GenericSystemTypeFreeJoint` class instead

mstoelzle commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,396 @@
+        pass
+
+
+class GenericSystemTypeFixedJoint(GenericSystemTypeFreeJoint, FixedJoint):
```
@skim0119 Thanks for your comment. I don't think your suggestion would work, as we need to access the `apply_torques` method of the `FixedJoint` class. We also need to inherit from `GenericSystemTypeFreeJoint` to access their `apply_forces` method. This `GenericSystemTypeFixedJoint` class is basically a combination of `GenericSystemTypeFreeJoint` and `FixedJoint`.

Please let me know if you have another idea how to solve this.

mstoelzle replied:
> I understand your point, but I feel it's little dangerous using multiple inheritances here. Those modules are not designed to be used as a mixin, and it can cause issues downstream.
> 
> For now, if you just want to assign the same method, simply put
> 
> ```python
> apply_torques = FixedJoint.apply_toruqes
> apply_forces = GenericSystemTypeFreeJoint.apply_forces
> ```
> 
> inside the class, instead of inheriting. You can also call `__init__` without inheriting, so your `__init__` method doesn't have to be changed.
> 
> > As you can see, the inheritance doesn't mean anything in your current implementation: it has a chance to be misinterpreted in later usage. Directly calling `GenericSystemTypeFreeJoint` and `FixedJoint` doesn't mean you are calling the parent entities; they are just calling other class members as static.
> 
> It appears to me that this class will eventually be converted to `FixedJoint` in the future, so let's keep the code as intuitive as possible.

This doesn't work, as the `__init__` cannot be called without inheriting. For the code
```py
FixedJoint.__init__(
            self, k=k, nu=nu, kt=kt, nut=nut, rest_rotation_matrix=rest_rotation_matrix
        )
```
I get the following error:
```
Traceback (most recent call last):
  File "/Users/maximilianstoelzle/Documents/phd/PyElastica/elastica/wrappers/connections.py", line 278, in __call__
    return self._connect_cls(*self._args, **self._kwargs)
  File "/Users/maximilianstoelzle/Documents/phd/PyElastica/elastica/experimental/connection_contact_joint/generic_system_type_connection.py", line 318, in __init__
    FixedJoint.__init__(
  File "/Users/maximilianstoelzle/Documents/phd/PyElastica/elastica/joint.py", line 240, in __init__
    super().__init__(k, nu)
TypeError: super(type, obj): obj must be an instance or subtype of type

```

mstoelzle replied:
@skim0119  In my mind we have four options:

1. You know of a solution how to call the (__init__) methods if non-inherited classes 
2. We leave the inheritance like it is now. I think the syntax is quite clear how we leverage the functionality of both base classes we inherit from. I don't think it will cause any issues as a) it is an experimental feature anyway, so interface and functionality changes need to be expected b) As you mentioned the code should hopefully be merged / replace the FixedJoint class. 
3. We just copy the necessary code from the FixedJoint class into the GenericSystemTypeFixedJoint class. 
4. We created a static method in the FixedJoint class, which can also be called from the GenericSystemTypeJoint class without inheriting.

@skim0119 Please let me know which way you want to go.

skim0119 replied:
@mstoelzle 
I guess its a bit tricky. For now, let's just copy the code for `__init__`. I'll clear the code when we merge to `joint.py`. In the end, I think the sequence needs to be modularized.

For `apply_force` and `apply_torques`, I think it shouldn't cause an issue, right?

## Suggestions:

- Use `FreeJoint` as a base
- Have its own `__init__` function
- Make alias for `apply_force` and `apply_torques`

mstoelzle replied:
> @mstoelzle I guess its a bit tricky. For now, let's just copy the code for `__init__`. I'll clear the code when we merge to `joint.py`. In the end, I think the initial computation needs to be modularized.
> 
> For `apply_force` and `apply_torques`, I think it shouldn't cause an issue, right?
> 
> ## Suggestions:
> * Use `FreeJoint` as a base
> * Have its own `__init__` function
> * Make alias for `apply_force` and `apply_torques`

@skim0119 I removed the inheritance from `FixedJoint` in https://github.com/GazzolaLab/PyElastica/pull/149/commits/80d0ed1511d41f457b4f6eb5994f6d0a3f8b661e. The `GenericSystemTypeFixedJoint` now uses `GenericSystemTypeFreeJoint` as a base class so that we can use their `__init__` without copying the code. The necessary `__init__` functionality from `FixedJoint` is copied.

mstoelzle commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,396 @@
+        relative_velocity = velocity_system_two - velocity_system_one
+
+        # Compute damping force considering the specified damping coefficient `nu`
+        damping_force = self.nu * relative_velocity
```
@skim0119 I actually looked into this topic yesterday again, but I would love to verify my thoughts with you guys:

We can formulate the problem such that we look at system one and consider the state of system two the desired (kinematic) goal state for system one. Based on the common Robotics literature such as  the book by Siciliano, we formulate our kinematic PD regulator as:

f = k_p (x_des - x) + k_d (v_des - v)

Where k_p and k_d are the proportional and damping gains respectively, x_des the desired position and v_des the desired velocity. 

Now we can adapt this formulation to our problem, as explained above:

f_1 = k (x_2 - x_1) + nu (v_2 - v_1)

This force f_1 will guide system one towards system two. 
Then, we apply the opposite force to system two to guide it towards the state of system one.

What I found though when running the spherical joint example is that this damping constant nu cannot be chosen too big, as the regulator becomes unstable. I assume that this comes from the rapidly changing "goal" state.

bhosale2 replied:
@mstoelzle thanks for explaining this, we actually noticed some inconsistencies on our end (described in #164). Although can you please set some small value of nu for the joints and confirm if it runs as expected? Maybe with a stabilized behavior? You don't need to set a high value.

mstoelzle replied:
@bhosale2 I haven't necessary verified how the damping verifies the multi-body system characteristics (e.g. whether it actually removes energy from the system), but I tested that the `generic_system_type_spherical_joint` example works without issues with a `k=1e3` and `nu=1e-1`. I didn't invest too much time on tuning these two values, but at some point when you increase either `k` or `nu` too much, the system becomes unstable.

bhosale2 replied:
Ok that behavior is expected thanks @mstoelzle

skim0119 commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,421 @@
+        nu: float,
+        kt: float,
+        nut: float = 0.0,
+        point_system_one: np.ndarray = None,
```
Add `Optional`. Also, please don't put the default value here, the default value should be described below (as you did 👍 . It's for documentation esthetic, so please understand.)

Optional already implies the default is None.

mstoelzle replied:
Python does not automatically initialize the parameter to the default value as hinted by the typing specification, see https://stackoverflow.com/questions/51710037/how-should-i-use-the-optional-type-hint

That is why we still need to set the default value to None, similar to https://github.com/GazzolaLab/PyElastica/blob/master/elastica/rod/factory_function.py#L23

mstoelzle replied:
Also, starting at Python 3.9 the `Optional` typing hint is not necessary anymore, see https://stackoverflow.com/a/67370289/1745305

mstoelzle commented on elastica/experimental/connection_contact_joint/generic_system_type_connection.py:
```diff
@@ -0,0 +1,421 @@
+        k: float,
+        nu: float,
+        point_system_one: np.ndarray = None,
+        point_system_two: np.ndarray = None,
```
Python does not automatically initialize the parameter to the default value as hinted by the typing specification, see https://stackoverflow.com/questions/51710037/how-should-i-use-the-optional-type-hint

That is why we still need to set the default value to None, similar to https://github.com/GazzolaLab/PyElastica/blob/master/elastica/rod/factory_function.py#L23

skim0119 reviewed (approved):
Looks good. I haven't run the test and examples, but we'll check them before releasing v0.3
